packages/imperative/src/config/src/ProfileInfo.ts

```diff
@@ -1225,6 +1234,312 @@ export class ProfileInfo {
+     */
+    public buildSchema(sources?: string[], layer?: IConfigLayer): IConfigSchema {
+        const finalSchema: Record<string, IProfileSchema> = {};
+        const desiredLayer = layer ?? this.getTeamConfig().layerActive();
+
+        const profileTypesInLayer = [...this.mProfileSchemaCache.entries()]
+            .filter(([type, _schema]) => type.includes(`${desiredLayer.path}:`));
+        for (const [typeWithPath, schema] of profileTypesInLayer) {
+            const type = typeWithPath.split(":").pop();
+            if (type == null) {
+                continue;
+            }
+
+            finalSchema[type] = schema;
+        }
+
+        let schemaEntries = Object.entries(finalSchema);
+        if (sources?.length > 0) {
+            schemaEntries = schemaEntries.filter(([typ, sch]) => {
+                if (!(typ in this.mExtendersJson.profileTypes)) {
+                    return false;
+                }
+
+                // If a list of sources were provided, ensure the type is contributed by at least one of these sources
+                if (sources.some((val) => this.mExtendersJson.profileTypes[typ].from.includes(val))) {
+                    return true;
+                }
+
+                return false;
+            });
+        }
+
+        return ConfigSchema.buildSchema(schemaEntries.map(([type, schema]) => ({
+            type,
+            schema
+        })));
+    }
+
+    /**
+     * @param [sources] (optional) Only include available types from the given list of sources
+     * @returns a list of all available profile types
+    */
+    public getProfileTypes(sources?: string[]): string[] {
+        const filteredBySource = sources?.length > 0;
+        const profileTypes = new Set<string>();
+        for (const layer of this.getTeamConfig().mLayers) {
+            if (layer.properties.$schema == null) continue;
+            const schemaUri = new url.URL(layer.properties.$schema, url.pathToFileURL(layer.path));
+            const schemaPath = url.fileURLToPath(schemaUri);
+            if (!fs.existsSync(schemaPath)) continue;
+
+            const profileTypesInLayer = [...this.mProfileSchemaCache.keys()].filter((key) => key.includes(`${layer.path}:`));
+            for (const typeWithPath of profileTypesInLayer) {
+                const type = typeWithPath.split(":").pop();
+                if (type == null) {
+                    continue;
+                }
+
+                profileTypes.add(type);
+            }
+        }
+
+        // Include all profile types from extenders.json if we are not filtering by source
+        if (filteredBySource) {
+            return [...profileTypes].filter((t) => {
+                if (!(t in this.mExtendersJson.profileTypes)) {
+                    return false;
+                }
+
+                return this.mExtendersJson.profileTypes[t].from.some((src) => sources.includes(src));
+            }).sort((a, b) => a.localeCompare(b));
```

Comment:
Considering the case of an invalid schema (i.e. new profile type not yet registered)...
I do think it's the correct thing to return an empty array for this.

However, should we have some extra validation and throw an error if any profileTypes in the provided source list is invalid?

Reply:
Discussed offline - will file a separate issue to validate schemas when grabbing them in `loadAllSchemas` 👍